### PR TITLE
feat(Backups): AWS Backups integration

### PR DESCRIPTION
**Pull request by izo3**
# Goal

Allow use of managed AWS Backups service

**To setup a sample backup you can use the following configuration:** 

```
import { Construct } from 'constructs';
import { App, TerraformStack } from 'cdktf';
import { AwsProvider } from '@cdktf/provider-aws';
import { ApplicationBackup } from './base/ApplicationBackups';

class Example extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);

    new AwsProvider(this, 'aws', {
      region: '<region>',
    });

    new ApplicationBackup(this, '<vault_name>', {
      name: 'Vault',
      kmsKeyArn: 'arn:aws:kms:<region>::<key_name>',
      prefix: '<prefix>',
      accountId: '<account_id>',
      vaultPolicy: '{"Version":"2012-10-17","Statement":[{"Sid":"Allowaccounttocopyintobackupvault","Effect":"Allow","Principal":{"AWS":"arn:aws:iam::<backup_account_id>:<backup_user>"},"Action":"backup:CopyIntoBackupVault","Resource":"*"}]}',
      backupPlans: [
        {
          name: 'Backups',
          resources: ['*'],
          selectionTag: [{
            key: '<selection_key>',
            type: 'STRINGEQUALS',
            value: '<selection_value>'
          }],
          rules: [
            {
              ruleName: 'DailyBackups',
              enableContinuousBackup: false,
              schedule: 'cron( 20 21 ? * * *)',
              lifecycle: {
                coldStorageAfter: 0,
                deleteAfter: 3
              },
              copyAction: [{
                destinationVaultArn: 'arn:aws:backup:<region>:<backup_account>:backup-vault:<vault_name>',
                lifecycle: {
                  coldStorageAfter: 0,
                  deleteAfter: 3
                }
              }

              ]
            },
          ],
        },
      ],
      tags: { service: '<service_name>',
              environment: '<environment>' },
    });
  }
}

const app = new App();
new Example(app, 'acme-example');
app.synth();
```

In the above example you need to substitute the following:

```
<region>  - Region of the aws accounts for the specific operations
<vault_name> - Name of the vault 
<prefix>  - Prefix attached to backup resources
<account_id> - Primary aws account id 
<backup_account_id>  - Backup aws account id, where backups will be copied
<backup_user>  - iam user authorized to copy backups
<key_name> - KMS key
<selection_key> - If key found in resources value will be inspected
<selection_value>  - If value matched backup will be triggered
```

### CODE REVIEW

File: src/base/ApplicationBackups.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    }[];
+    tags?: { [key: string]: string };
+  }
+
+export class ApplicationBackup extends Resource {    
+
+    public backupPlan: Backup.BackupPlan;
+    public backupSelection: Backup.BackupSelection;
+    public backupPlanRule: Backup.BackupPlanRule;
+    // private static vault: Backup.BackupVault=null;
```
Comment (kkyeboah) [inline src/base/ApplicationBackups.ts:32]:
```suggestion
```

File: src/base/ApplicationBackups.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    private static vault: Backup.BackupVault;
+    // public readonly config: ApplicationBackupProps;
```
Comment (kkyeboah) [inline src/base/ApplicationBackups.ts:34]:
```suggestion
```

File: src/base/ApplicationBackups.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    name: string;
+    kmsKeyArn: string;
+    prefix: string;
+    accountId: string;
+    vaultPolicy: string;
+    backupPlans: {
+      name: string;
+      resources: string[];
+      rules: Omit<BackupPlanRule, 'targetVaultName'>[];
+      // selectionTag?: Omit<BackupPlanRule, 'targetVaultName'>[];
```
Comment (kkyeboah) [inline src/base/ApplicationBackups.ts:21]: We should find the correct type for this property

**Review by kkyeboah — COMMENTED** [review]
Construct looks good to me. We have a number of linter errors though

File: src/base/ApplicationBackups.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+              
+        const vault = new BackupVault(this, 'backup-vault', {
+            name: `${config.prefix}-${config.name}`,
+            kmsKeyArn: config.kmsKeyArn,
+            tags: config.tags,
+        })
+        
+        const vaultPolicy = new BackupVaultPolicy(this, 'backup-vault-policy', {
+            backupVaultName: `${config.prefix}-${config.name}`,
+            policy: config.vaultPolicy
```
Comment (kkyeboah) [inline src/base/ApplicationBackups.ts:51]: is there a default policy we could use here?

**Comment by bkochendorfer** [comment]
@izo3 this is marked as a draft are you still working on it or are you ready for review?

**Comment by izo3** [comment]
I am moving this PR to this one https://github.com/Pocket/terraform-modules/pull/902

**Pull request closed without merging**